cirq/value/states.py

```diff
@@ -0,0 +1,365 @@
+        return self.on(*args, **kwargs)
+
+    @abc.abstractmethod
+    def state_vector(self):
+        """Return a state vector representation of the named state."""
+
+    def projector(self):
+        """Return |s><s| as a matrix for the named state."""
+        vec = self.state_vector()[:, np.newaxis]
+        return vec @ vec.conj().T
+
+    def _json_dict_(self):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return protocols.obj_to_dict_helper(self, [])
+
+    def __eq__(self, other):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return self.__class__ == other.__class__
+
+    def __hash__(self):
+        # Descendants should be singletons determined solely by the class name.
+        # Otherwise, you must override this method.
+        return hash(self.__class__.__name__)
+
+
+@dataclass(frozen=True)
+class TensorProductState:
+    """A quantum state that is a tensor product of one qubit states.
+
+    For example, the |00> state is `cirq.PLUS_Z(q0) * cirq.PLUS_Z(q1)`.
+    The |+> state is a length-1 tensor product state and can be constructed
+    with `cirq.PLUS_X(q0)`.
+    """
+    states: Dict['cirq.Qid', _Named1QState]
+
+    def __init__(self, states=None):
+        if states is None:
+            # coverage: ignore
+            states = dict()
+
+        object.__setattr__(self, 'states', states)
+
+    @property
+    def qubits(self):
+        return sorted(self.states.keys())
+
+    def __mul__(self, other):
+        if not isinstance(other, TensorProductState):
+            raise ValueError("Multiplication is only supported "
+                             "with other TensorProductStates.")
+
+        dupe_qubits = set(other.states.keys()) & set(self.states.keys())
+        if len(dupe_qubits) != 0:
+            raise ValueError("You tried to multiply two states, "
+                             "but both contain factors for these qubits: {}"
+                             .format(sorted(dupe_qubits)))
+
+        new_states = self.states
+        new_states.update(other.states)
+        return TensorProductState(new_states)
+
+    def __str__(self):
+        return ' * '.join(f'{st}({q})' for q, st in self.states.items())
+
+    def __repr__(self):
+        states_dict_repr = ', '.join(f'{repr(key)}: {repr(val)}'
+                                     for key, val in self.states.items())
+        return 'cirq.TensorProductState({%s})' % states_dict_repr
+
+    def __getitem__(self, qubit):
+        """Return the NamedState at the given qubit."""
+        return self.states[qubit]
+
+    def __iter__(self):
+        yield from self.states.items()
+
+    def __len__(self):
+        return len(self.states)
+
+    def __eq__(self, other):
+        if not isinstance(other, TensorProductState):
+            return False
```

Comment:
return NotImplemented instead of False, to allow later classes to override the behavior. If you use `cirq.testing.EqualsTester` it should catch this.